numba/np/ufunc/deviceufunc.py

```diff
@@ -732,110 +724,164 @@ def _broadcast_scalar_input(self, ary, shape):
+    is specific to the arguments given to the GUFunc call.
+
+    The base class implements the overall logic; subclasses provide
+    target-specific implementations of individual functions.
+    """
+
+    # The base class uses these slots; subclasses may provide additional slots.
     __slots__ = [
-        'args',
-        'kwargs',
-        'output',
-        'norm_inputs',
-        'kernel_returnvalue',
-        'kernel_parameters',
-        '_is_device_array',
-        '_need_device_conversion',
+        'outputs',
+        'inputs',
+        '_copy_result_to_host',
     ]
 
+    @abstractmethod
+    def launch_kernel(self, kernel, nelem, args):
+        """Implement the kernel launch"""
+
+    @abstractmethod
+    def is_device_array(self, obj):
+        """
+        Return True if `obj` is a device array for this target
+        """
+
+    @abstractmethod
+    def as_device_array(self, obj):
+        """
+        Return `obj` as a device array on this target.
+
+        May return `obj` directly if it is already on the target.
+        """
+
+    @abstractmethod
+    def to_device(self, hostary):
+        """
+        Copy `hostary` to the device and return the device array.
+        """
+
+    @abstractmethod
+    def device_array(self, shape, dtype):
+        """
+        Allocate a new uninitialized device array with the given shape and
+        dtype.
+        """
+
     def __init__(self, nin, nout, args, kwargs):
-        if nout > 1:
-            raise ValueError('multiple output is not supported')
-        self.args = args
-        self.kwargs = kwargs
-
-        user_output_is_device = False
-        self.output = self.kwargs.get('out')
-        if self.output is not None:
-            user_output_is_device = self.is_device_array(self.output)
-            if user_output_is_device:
-                self.output = self.as_device_array(self.output)
-        self._is_device_array = [self.is_device_array(a) for a in self.args]
-        self._need_device_conversion = (not any(self._is_device_array) and
-                                        not user_output_is_device)
-
-        # Normalize inputs
-        inputs = []
-        for a, isdev in zip(self.args, self._is_device_array):
-            if isdev:
-                inputs.append(self.as_device_array(a))
+        # Get outputs either from the out kwarg, or generate an initial list of
+        # "placeholder" outputs using None as a sentry value
+        outputs = kwargs.get('out')
+        if outputs is not None and len(args) > nin:
+            raise ValueError("cannot specify 'out' as both positional "
+                             "and keyword arguments")
+        else:
+            outputs = [outputs] * nout
+
+        # Ensure all output device arrays are Numba device arrays - any output
+        # passed in that supports the CUDA Array Interface is converted; others
+        # are left untouched.
```

Comment:
Is this CUDA specific in a generalized implementation? (seems like it may just be the comment referencing CUDA opposed to the implementation).

Reply:
I think the implementation here is generalized, in that if there were a "ROC Array Interface" then a ROC target would also be able to convert its arguments to "Numba ROC device arrays", but there only happens to be the CUDA implementation at present. I feel that this comment could be fixed by adding a "for example" to it - I'd like to retain it here as it does explain what's going on in the following lines.

Reply:
Now changed so the comment explains that it's just an example of a scenario rather than the sole reason for this functionality here.

Reply:
Thanks, looks good.